File: update_db_from_config.py
#!/usr/bin/env python3
"""Update database with repository information from config.yaml"""

import sys
from pathlib import Path

import yaml

# Add project to path
sys.path.insert(0, str(Path(__file__).parent))

from merge_god.db_operations import StateDatabase  # noqa: E402


def main():
    """Load config and update database"""
    config_path = Path("config.yaml")
    db_path = Path("merge-god-state.db")

    if not config_path.exists():
        print(f"❌ Config file not found: {config_path}")
        return 1

    print(f"Loading config from: {config_path}")
    with open(config_path) as f:
        config = yaml.safe_load(f)

    if not config or "repos" not in config:
        print("❌ Invalid config: missing 'repos' section")
        return 1

    print(f"Updating database: {db_path}")
    db = StateDatabase(db_path)

    # Process each repository in config
    for repo_config in config["repos"]:
        if not repo_config.get("enabled", True):
            print(f"  Skipping disabled repo: {repo_config.get('name', 'unknown')}")
            continue

        repo_path = Path(repo_config["path"])
        repo_name = repo_config.get("name", repo_path.name)

        if not repo_path.exists():
            print(f"  ⚠️  Repository path not found: {repo_path}")
            continue

        # Get default branch (try main first, fallback to master)
        default_branch = "main"  # Could query git here if needed

        print(f"  Adding repository: {repo_name}")
        print(f"    Path: {repo_path}")
        print(f"    Default branch: {default_branch}")

        db.save_repository(name=repo_name, path=str(repo_path), default_branch=default_branch)

    print("\n✅ Database updated successfully!")

    # Show statistics
    stats = db.get_statistics()
    print("\nDatabase statistics:")
    print(f"  Repositories: {stats['repositories']}")
    print(f"  PR snapshots: {stats['pr_snapshots']}")
    print(f"  Processing records: {stats['processing_records']}")
    print(f"  Database size: {stats['database_size_bytes']} bytes")

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: update_db_from_config.ts
#!/usr/bin/env node
/**
 * Update database with repository information from config.yaml.
 *
 * Ported from update_db_from_config.py.
 */

import { existsSync, readFileSync } from "node:fs";
import path from "node:path";
import { pathToFileURL } from "node:url";

import YAML from "yaml";

import { SyncStore } from "@merge-god/github-sync";

async function main(): Promise<number> {
  const configPath = "config.yaml";
  const dbPath = "merge-god-state.db";

  if (!existsSync(configPath)) {
    console.log(`❌ Config file not found: ${configPath}`);
    return 1;
  }

  console.log(`Loading config from: ${configPath}`);
  const parsed: unknown = YAML.parse(readFileSync(configPath, "utf8"));

  if (!parsed || typeof parsed !== "object" || !("repos" in parsed)) {
    console.log("❌ Invalid config: missing 'repos' section");
    return 1;
  }

  const repos = (parsed as Record<string, unknown>)["repos"];
  if (!Array.isArray(repos)) {
    console.log("❌ Invalid config: 'repos' is not a list");
    return 1;
  }

  console.log(`Updating database: ${dbPath}`);
  const db = new SyncStore(dbPath);
  await db.initialize();

  for (const item of repos) {
    const repoConfig =
      item && typeof item === "object" ? (item as Record<string, unknown>) : {};

    const enabledRaw = repoConfig["enabled"];
    const enabled = enabledRaw === undefined ? true : Boolean(enabledRaw);
    if (!enabled) {
      const name = (repoConfig["name"] as string | undefined) ?? "unknown";
      console.log(`  Skipping disabled repo: ${name}`);
      continue;
    }

    const repoPath = repoConfig["path"];
    if (typeof repoPath !== "string") {
      console.log("  ⚠️  Repository missing valid 'path'; skipping");
      continue;
    }

    const repoName = (repoConfig["name"] as string | undefined) ?? path.basename(repoPath);

    if (!existsSync(repoPath)) {
      console.log(`  ⚠️  Repository path not found: ${repoPath}`);
      continue;
    }

    const defaultBranch = "main";

    console.log(`  Adding repository: ${repoName}`);
    console.log(`    Path: ${repoPath}`);
    console.log(`    Default branch: ${defaultBranch}`);

    await db.saveRepository(repoName, repoPath, defaultBranch);
  }

  console.log("\n✅ Database updated successfully!");

  const stats = (await db.getStatistics()) as Record<string, unknown>;
  console.log("\nDatabase statistics:");
  for (const [key, value] of Object.entries(stats)) {
    console.log(`  ${key}: ${value}`);
  }

  db.close();
  return 0;
}

if (import.meta.url === pathToFileURL(process.argv[1] ?? "").href) {
  void main().then((code) => process.exit(code));
}
